### ISSUE TITLE
Race condition in pre-start script?

### ISSUE THREAD

**Issue by peterellisjones** [cloudfoundry/capi-release#147]
Hi capi-release team,

When deploying a new foundation we get the following error in the cloud_controller_ng pre-start script (logs from `/var/vcap/sys/log/cloud_controller_ng/pre-start.stderr.log`):

```
+ export LANG=en_US.UTF-8
+ LANG=en_US.UTF-8
+ source /var/vcap/packages/capi_utils/syslog_utils.sh
++ basename /var/vcap/jobs/cloud_controller_ng/bin/pre-start
+ tee_output_to_sys_log cloud_controller_ng.pre-start
+ declare log_name=cloud_controller_ng.pre-start
+ exec
+ exec
++ tee -a /dev/fd/63
++ prepend_datetime
++ awk -W interactive '{ system("echo -n [$(date +\"%Y-%m-%d %H:%M:%S%z\")]"); print " " $0 }'
+++ logger -p user.info -t vcap.cloud_controller_ng.pre-start.stdout
++ prepend_datetime
++ awk -W interactive '{ system("echo -n [$(date +\"%Y-%m-%d %H:%M:%S%z\")]"); print " " $0 }'
++ tee -a /dev/fd/63
+++ logger -p user.error -t vcap.cloud_controller_ng.pre-start.stderr
[2019-06-27 15:19:31+0000] ++ dirname /var/vcap/jobs/cloud_controller_ng/bin/pre-start
[2019-06-27 15:19:31+0000] + SCRIPT_DIR=/var/vcap/jobs/cloud_controller_ng/bin
[2019-06-27 15:19:31+0000] + CC_JOB_DIR=/var/vcap/jobs/cloud_controller_ng
[2019-06-27 15:19:31+0000] + CC_PACKAGE_DIR=/var/vcap/packages/cloud_controller_ng
[2019-06-27 15:19:31+0000] + CONFIG_DIR=/var/vcap/jobs/cloud_controller_ng/config
[2019-06-27 15:19:31+0000] + export CLOUD_CONTROLLER_NG_CONFIG=/var/vcap/jobs/cloud_controller_ng/config/cloud_controller_ng.yml
[2019-06-27 15:19:31+0000] + CLOUD_CONTROLLER_NG_CONFIG=/var/vcap/jobs/cloud_controller_ng/config/cloud_controller_ng.yml
[2019-06-27 15:19:31+0000] + export BUNDLE_GEMFILE=/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/Gemfile
[2019-06-27 15:19:31+0000] + BUNDLE_GEMFILE=/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/Gemfile
[2019-06-27 15:19:31+0000] + source /var/vcap/jobs/cloud_controller_ng/bin/setup_local_blobstore.sh
[2019-06-27 15:19:31+0000] + source /var/vcap/jobs/cloud_controller_ng/bin/ruby_version.sh
[2019-06-27 15:19:31+0000] ++ source /var/vcap/packages/ruby-2.5.5-r0.10.0/bosh/runtime.env
[2019-06-27 15:19:31+0000] +++ export PATH=/var/vcap/packages/ruby-2.5.5-r0.10.0/bin:/usr/sbin:/usr/bin:/sbin:/bin
[2019-06-27 15:19:31+0000] +++ PATH=/var/vcap/packages/ruby-2.5.5-r0.10.0/bin:/usr/sbin:/usr/bin:/sbin:/bin
[2019-06-27 15:19:31+0000] +++ export GEMRC=/var/vcap/packages/ruby-2.5.5-r0.10.0/bosh/gemrc
[2019-06-27 15:19:31+0000] +++ GEMRC=/var/vcap/packages/ruby-2.5.5-r0.10.0/bosh/gemrc
[2019-06-27 15:19:31+0000] + main
[2019-06-27 15:19:31+0000] + start_bosh_dns_or_consul
[2019-06-27 15:19:31+0000] + '[' -f /var/vcap/jobs/bosh-dns/bin/is-system-resolver ']'
[2019-06-27 15:19:31+0000] + /var/vcap/jobs/bosh-dns/bin/is-system-resolver
[2019-06-27 15:19:31+0000] + echo 'waiting for bosh_dns'
[2019-06-27 15:19:31+0000] + /var/vcap/jobs/bosh-dns/bin/wait
[2019-06-27 15:19:36+0000] + setup_directories
[2019-06-27 15:19:36+0000] + setup_local_blobstore
[2019-06-27 15:19:36+0000] + :
[2019-06-27 15:19:36+0000] + setup_api_directories
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/sys/run/cloud_controller_ng
[2019-06-27 15:19:36+0000] + chown -R vcap:vcap /var/vcap/sys/run/cloud_controller_ng
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/sys/log/cloud_controller_ng
[2019-06-27 15:19:36+0000] + chown -R vcap:vcap /var/vcap/sys/log/cloud_controller_ng
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/data/cloud_controller_ng/tmp
[2019-06-27 15:19:36+0000] + chown vcap:vcap /var/vcap/data/cloud_controller_ng/tmp
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/sys/cores
[2019-06-27 15:19:36+0000] + chown vcap:vcap /var/vcap/sys/cores
[2019-06-27 15:19:36+0000] + BUNDLER_DIR1=/tmp/bundler
[2019-06-27 15:19:36+0000] + BUNDLER_DIR2=/var/vcap/data/cloud_controller_ng/tmp/bundler
[2019-06-27 15:19:36+0000] + chpst -u vcap:vcap mkdir -p /tmp/bundler /var/vcap/data/cloud_controller_ng/tmp/bundler
[2019-06-27 15:19:36+0000] + chpst -u vcap:vcap chmod -R go-w /tmp/bundler /var/vcap/data/cloud_controller_ng/tmp/bundler
[2019-06-27 15:19:36+0000] + echo /var/vcap/sys/cores/core-%e-%s-%p-%t
[2019-06-27 15:19:36+0000] + setup_nginx_directories
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/sys/run/nginx_cc
[2019-06-27 15:19:36+0000] + chown -R vcap:vcap /var/vcap/sys/run/nginx_cc
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/sys/log/nginx_cc
[2019-06-27 15:19:36+0000] + chown -R vcap:vcap /var/vcap/sys/log/nginx_cc
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/data/cloud_controller_ng/tmp/uploads
[2019-06-27 15:19:36+0000] + chown vcap:vcap /var/vcap/data/cloud_controller_ng/tmp/uploads
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/data/cloud_controller_ng/tmp/staged_droplet_uploads
[2019-06-27 15:19:36+0000] + chown vcap:vcap /var/vcap/data/cloud_controller_ng/tmp/staged_droplet_uploads
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/data/nginx_cc/tmp
[2019-06-27 15:19:36+0000] + chown vcap:vcap /var/vcap/data/nginx_cc/tmp
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/packages/nginx/proxy_temp
[2019-06-27 15:19:36+0000] + touch /var/vcap/packages/nginx/logs/error.log
[2019-06-27 15:19:36+0000] + chown vcap:vcap /var/vcap/packages/nginx/logs/error.log
[2019-06-27 15:19:36+0000] + setup_nginx_plugin_directories
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/sys/run/nginx_newrelic_plugin
[2019-06-27 15:19:36+0000] + chown -R vcap:vcap /var/vcap/sys/run/nginx_newrelic_plugin
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/sys/log/nginx_newrelic_plugin
[2019-06-27 15:19:36+0000] + chown -R vcap:vcap /var/vcap/sys/log/nginx_newrelic_plugin
[2019-06-27 15:19:36+0000] + cp /var/vcap/jobs/cloud_controller_ng/config/newrelic_plugin.yml /var/vcap/packages/nginx_newrelic_plugin/config
[2019-06-27 15:19:36+0000] + chown -R vcap:vcap /var/vcap/packages/nginx_newrelic_plugin/config/newrelic_plugin.yml
[2019-06-27 15:19:36+0000] + mkdir -p /var/vcap/data/cloud_controller_ng/diagnostics
[2019-06-27 15:19:36+0000] + chown vcap:vcap /var/vcap/data/cloud_controller_ng/diagnostics
[2019-06-27 15:19:36+0000] + perform_migration
[2019-06-27 15:19:36+0000] + pushd /var/vcap/packages/cloud_controller_ng/cloud_controller_ng
[2019-06-27 15:19:36+0000] + chpst -u vcap:vcap /var/vcap/jobs/cloud_controller_ng/bin/migrate_db
[2019-06-27 15:19:36+0000] + CC_PACKAGE_DIR=/var/vcap/packages/cloud_controller_ng
[2019-06-27 15:19:36+0000] + export BUNDLE_GEMFILE=/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/Gemfile
[2019-06-27 15:19:36+0000] + BUNDLE_GEMFILE=/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/Gemfile
[2019-06-27 15:19:36+0000] + export STACKS_YML=/var/vcap/jobs/cloud_controller_ng/config/stacks.yml
[2019-06-27 15:19:36+0000] + STACKS_YML=/var/vcap/jobs/cloud_controller_ng/config/stacks.yml
[2019-06-27 15:19:36+0000] + main
[2019-06-27 15:19:36+0000] + migrate
[2019-06-27 15:19:36+0000] + set +e
[2019-06-27 15:19:36+0000] + echo 'Running migrations'
[2019-06-27 15:19:36+0000] + local result=1
[2019-06-27 15:19:36+0000] + local counter=1
[2019-06-27 15:19:36+0000] + local num_retries=3
[2019-06-27 15:19:36+0000] + [[ 1 -le 3 ]]
[2019-06-27 15:19:36+0000] + echo 'Running migration try number 1 of 3'
[2019-06-27 15:19:36+0000] + bundle exec rake db:migrate
[2019-06-27 15:20:22+0000] + result=0
[2019-06-27 15:20:22+0000] + [[ 0 -eq 0 ]]
[2019-06-27 15:20:22+0000] + break
[2019-06-27 15:20:22+0000] + [[ 0 -ne 0 ]]
[2019-06-27 15:20:22+0000] + set -e
[2019-06-27 15:20:22+0000] + exit 0
[2019-06-27 15:20:22+0000] + popd
[2019-06-27 15:20:22+0000] + seed_db
[2019-06-27 15:20:22+0000] + pushd /var/vcap/packages/cloud_controller_ng/cloud_controller_ng
[2019-06-27 15:20:22+0000] + chpst -u vcap:vcap /var/vcap/jobs/cloud_controller_ng/bin/seed_db
[2019-06-27 15:20:33+0000] rake aborted!
[2019-06-27 15:20:33+0000] Errno::ECONNRESET: Connection reset by peer - SSL_connect
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient/ssl_socket.rb:103:in `connect'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient/ssl_socket.rb:103:in `ssl_connect'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient/ssl_socket.rb:41:in `initialize'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient/ssl_socket.rb:26:in `new'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient/ssl_socket.rb:26:in `create_socket'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient/session.rb:752:in `block in connect'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient/session.rb:748:in `connect'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient/session.rb:511:in `query'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient/session.rb:177:in `query'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient.rb:1242:in `do_get_block'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient.rb:1019:in `block in do_request'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient.rb:1133:in `protect_keep_alive_disconnected'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient.rb:1014:in `do_request'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient.rb:856:in `request'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/gem_home/ruby/2.5.0/gems/httpclient-2.8.3/lib/httpclient.rb:743:in `get'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/cloud_controller/routing_api/routing_api_client.rb:34:in `router_groups'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/cloud_controller/routing_api/routing_api_client.rb:48:in `router_group_guid'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/cloud_controller/seeds.rb:111:in `find_routing_guid'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/cloud_controller/seeds.rb:78:in `block in create_seed_domains'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/cloud_controller/seeds.rb:75:in `each'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/cloud_controller/seeds.rb:75:in `create_seed_domains'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/cloud_controller/seeds.rb:11:in `write_seed_data'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/tasks/db.rake:110:in `block (3 levels) in <top (required)>'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/cloud_controller/background_job_environment.rb:15:in `setup_environment'
[2019-06-27 15:20:33+0000] /var/vcap/data/packages/cloud_controller_ng/adca8e87a9d0822603620e2cda7e59e4dc6e594b/cloud_controller_ng/lib/tasks/db.rake:109:in `block (2 levels) in <top (required)>'
[2019-06-27 15:20:33+0000] /var/vcap/packages/cloud_controller_ng/gem_home/ruby/2.5.0/gems/rake-12.3.2/exe/rake:27:in `<top (required)>'
[2019-06-27 15:20:33+0000] /var/vcap/packages/ruby-2.5.5-r0.10.0/bin/bundle:30:in `block in <main>'
[2019-06-27 15:20:33+0000] /var/vcap/packages/ruby-2.5.5-r0.10.0/bin/bundle:22:in `<main>'
[2019-06-27 15:20:33+0000] Tasks: TOP => db:seed
[2019-06-27 15:20:33+0000] (See full trace by running task with --trace)
```

It is fixed by commenting out this line https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/seeds.rb#L11 in `/var/vcap/data/packages/cloud_controller_ng/$package_id/cloud_controller_ng/lib/cloud_controller/seeds.rb` on the first API instance that fails and running `bosh deploy` again. My best guess is that the task tries to make requests to the routing API which hasn't started yet (https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/seeds.rb#L108-L116)

**Comment by peterellisjones** [comment]
Some more information: we are using https://github.com/cloudfoundry/cf-deployment to build our manifest (currently v9.3.0), are deploying to Openstack and have 3 CF API instances

**Comment by tcdowney** [comment]
Thanks for reporting this @peterellisjones!

I feel like it doesn't _always_ call out to the routing API during seeding, but may be incorrect. @peterellisjones are you seeding a domain that is associated with a TCP router?

If we're always calling out to the routing API during seeding it would make sense to block in pre-start and wait for it to come up (like we do for the blobstore). @adobley as one of the Networking Program's PMs and someone more familiar with how the routing-api starts up during deployment, do you have any recommendations for us on how we should handle this?

**Comment by adobley** [comment]
@tcdowney in our [pre-start](https://github.com/cloudfoundry/routing-release/blob/develop/jobs/routing-api/templates/bpm-pre-start.erb) we wait until the DNS is up, by [default](https://github.com/cloudfoundry/routing-release/blob/develop/jobs/routing-api/spec#L104-L106) waiting for `uaa.service.cf.internal` to be resolvable. It might be taking a bit longer to start up the first time while it sets up its DB.

Is there an issue that it is reporting healthy before it is serving requests or that it is not available yet?

There isn't really a health-checking endpoint. What we do with policy server is to just do a `GET` to tell if the service we need to talk to is available or not.

cc/ @KauzClay @rodolfo2488

**Comment by peterellisjones** [comment]
> I feel like it doesn't always call out to the routing API during seeding, but may be incorrect. @peterellisjones are you seeding a domain that is associated with a TCP router?

Hi @tcdowney — Thanks for the quick response. Yes, here is our domains block 
```
  - name: cloud_controller_ng
    properties:
      app_domains:
      - ((cf_apps_domain))
      - internal: true
        name: apps.internal
      - name: tcp.((cf_apps_domain))
        router_group_name: default-tcp
```

**Comment by tcdowney** [comment]
Thanks @peterellisjones, that will help us reproduce and could explain why we haven't seen this in our team's CI environments (one has TCP routers, but I don't believe it seeds any domains for it).

**Comment by selzoc** [comment]
Looks like this was fixed in  c7f6a36 and 5bfc18d
